cpp/src/parquet/column_reader.cc

```diff
@@ -861,8 +864,113 @@ class TypedColumnReaderImpl : public TypedColumnReader<DType>,
   Type::type type() const override { return this->descr_->physical_type(); }
 
   const ColumnDescriptor* descr() const override { return this->descr_; }
+
+  ExposedEncodingType GetExposedEncoding() override { return this->exposed_encoding_; };
+
+  int64_t ReadBatchWithDictionary(int64_t batch_size, int16_t* def_levels,
+                                  int16_t* rep_levels, int32_t* indices,
+                                  int64_t* indices_read, const T** dict,
+                                  int32_t* dict_len) override;
+
+ protected:
+  void SetExposedEncoding(ExposedEncodingType encoding) override {
+    this->exposed_encoding_ = encoding;
+  }
+
+ private:
+  /// \brief Read dictionary indices. Similar to ReadValues but decode data to
+  /// dictionary indices.
+  ///
+  /// Called only by ReadBatchWithDictionary().
+  ///
+  /// \note API EXPERIMENTAL
+  int64_t ReadDictionaryIndices(int64_t indices_to_read, int32_t* indices) {
+    auto decoder = dynamic_cast<DictDecoder<DType>*>(this->current_decoder_);
```

Comment:
it would make sense to add DCHECK here to validate the expose mode is correct.

Also use checked_cast<> internal which does a IFDEF for debug/non debug mode that validates the type of the decoder.

Reply:
I see that you check below before calling this method. DCHECK for documentation purposes can still make sense.

Reply:
Thanks. Adding DCHECK may conflict with unit tests of column reader. The DCHECK assumes the column reader has enabled exposing dictionary encoding, which can only be enabled by a file reader object (via calling the protected SetExposedEncoding method). That means, the DCHECK would require a file reader object for the column reader's unit test. Since there's a checking before calling this method, maybe we can skip the DCHECK.